// @cosmograph/ui/modules/bars/index.js
var x = Object.defineProperty;
var L = (m, u, t) => u in m ? x(m, u, { enumerable: !0, configurable: !0, writable: !0, value: t }) : m[u] = t;
var n = (m, u, t) => L(m, typeof u != "symbol" ? u + "" : u, t);
import { defaultMaxDisplayedItems as M, defaultBarsConfig as p } from "./config.js";
import B from "./style.module.css.js";
import E from "../../styles.module.css.js";
import { merge as w } from "../../utils.js";
import { DisplayStateManager as b } from "../_abstract/display-state-manager/index.js";
class $ extends b {
  /**
   * Constructs a new instance of the Bars class, which is responsible for rendering a bar chart visualization.
   *
   * @param container - The HTML element that will contain the bar chart.
   * @param config - An optional configuration object that can be used to customize the behavior of the bar chart.
   */
  constructor(t, e) {
    const i = w(p, e ?? {}), s = b.extractConfig(i, p);
    super(t, s);
    n(this, "_data", []);
    n(this, "_originalData", []);
    n(this, "_config");
    n(this, "_container");
    n(this, "_searchContainer");
    n(this, "_searchInput");
    n(this, "_clearButton");
    n(this, "_scrollContainer");
    n(this, "_barsContainer");
    n(this, "_highlightedBarsContainer");
    n(this, "_showMoreButton");
    n(this, "_showMoreCount");
    n(this, "_sortingBlock");
    n(this, "_uniqueValues", /* @__PURE__ */ new Set());
    n(this, "_countMap", new Int32Array(0));
    n(this, "_visibleItemsCount", 0);
    n(this, "_hiddenItemsSum", 0);
    n(this, "_itemHeight", 0);
    n(this, "_itemMargin", 0);
    n(this, "_startIndex", 0);
    n(this, "_endIndex", 0);
    n(this, "_maxCount", 0);
    n(this, "_selectedItem");
    n(this, "_selectedLabels", /* @__PURE__ */ new Set());
    n(this, "_isExpanded", !1);
    n(this, "_sortingHandlers");
    n(this, "_highlightedCounts");
    n(this, "_isFilteringActive", !1);
    n(this, "_isConfigResortInProgress", !1);
    n(this, "_updateVisibleItems", () => {
      const t = this._itemHeight + this._itemMargin;
      this._startIndex = Math.floor(this._scrollContainer.scrollTop / t), this._endIndex = Math.min(this._startIndex + this._visibleItemsCount + 1, this._data.length), this._renderItems();
    });
    n(this, "_handleShowMoreCount", () => {
      var i, s, o, a, h, r;
      const t = this._data.length - this._maxDisplayedItems, e = ((s = (i = this._config).countFormatter) == null ? void 0 : s.call(i, this._hiddenItemsSum)) ?? this._hiddenItemsSum.toString();
      if (this._isFilteringActive && this._config.showTotalWhenFiltered) {
        const _ = this._data.slice(this._maxDisplayedItems).reduce((g, C) => g + C.count, 0), d = ((a = (o = this._config).countFormatter) == null ? void 0 : a.call(o, _)) ?? _.toString();
        this._showMoreCount.textContent = `${e} / ${d}`;
      } else
        this._showMoreCount.textContent = e;
      this._showMoreButton.firstChild.textContent = `${((r = (h = this._config).countFormatter) == null ? void 0 : r.call(h, t)) ?? t.toString()} more`;
    });
    n(this, "_handleScroll", () => {
      this._updateVisibleItems();
    });
    n(this, "_handleSort", () => {
      this._config.sort === "alphabetical" ? this.sortAlphabetically(this._config.sortAscending) : this.sortByCount(this._config.sortAscending);
    });
    n(this, "_handleSearch", () => {
      const t = this._searchInput.value.toLowerCase();
      this._clearButton.classList[t ? "add" : "remove"]("visible"), this._data = this._originalData.filter((e) => String(e.label).toLowerCase().includes(t)), this._handleSort();
    });
    n(this, "_clearSearch", () => {
      this._searchInput.value = "", this._handleSearch();
    });
    n(this, "_handleShowMore", () => {
      this._isExpanded = !this._isExpanded, this._isExpanded ? (this._scrollContainer.style.overflowY = "auto", this._showMoreButton.firstChild.textContent = "show less", this._showMoreCount.textContent = "") : (this._scrollContainer.style.overflowY = "hidden", this._scrollContainer.scrollTop = 0, this._handleShowMoreCount());
    });
    this._config = i, this._container = t, this._container.classList.add(B.bars), this._container.classList.add(E.scrollbar), this._scrollContainer = document.createElement("div"), this._scrollContainer.classList.add("scroll-container"), this._barsContainer = document.createElement("div"), this._barsContainer.classList.add("bars-container"), this._highlightedBarsContainer = document.createElement("div"), this._highlightedBarsContainer.classList.add("bars-container", "highlighted-bars-container"), this._scrollContainer.appendChild(this._barsContainer), this._scrollContainer.appendChild(this._highlightedBarsContainer), this._showMoreButton = this._createShowMoreButton(), this._sortingBlock = this._createSortingBlock(), this._searchContainer = this._createSearch(), this._container.appendChild(this._searchContainer), this._container.appendChild(this._scrollContainer), this._container.appendChild(this._showMoreButton), this._container.appendChild(this._sortingBlock), this._scrollContainer.addEventListener("scroll", this._handleScroll, { passive: !0 });
  }
  _isSortAnimationEnabled() {
    var t, e;
    return this._config.animateSort === !1 ? !1 : !((e = (t = window.matchMedia) == null ? void 0 : t.call(window, "(prefers-reduced-motion: reduce)")) != null && e.matches);
  }
  _captureSortPositions() {
    const t = /* @__PURE__ */ new Map();
    for (const i of Array.from(this._barsContainer.querySelectorAll(".row"))) {
      const s = i.dataset.label;
      s && t.set(s, i.offsetTop);
    }
    const e = /* @__PURE__ */ new Map();
    for (const i of Array.from(this._highlightedBarsContainer.querySelectorAll(".bar-highlighted"))) {
      const s = i.dataset.label;
      s && e.set(s, i.offsetTop);
    }
    return { base: t, highlighted: e };
  }
  _animateSortFrom(t) {
    const e = this._config.sortAnimationDurationMs ?? 250, i = "ease-out", s = (o, a, h) => {
      for (const r of Array.from(o.querySelectorAll(a))) {
        const l = r.dataset.label;
        if (!l)
          continue;
        const _ = h.get(l);
        if (_ === void 0)
          continue;
        const d = _ - r.offsetTop;
        if (d) {
          for (const g of r.getAnimations())
            g.cancel();
          r.animate([{ transform: `translateY(${d}px)` }, { transform: "translateY(0)" }], { duration: e, easing: i });
        }
      }
    };
    requestAnimationFrame(() => {
      s(this._barsContainer, ".row", t.base), s(this._highlightedBarsContainer, ".bar-highlighted", t.highlighted);
    });
  }
  _getFilteredCount(t) {
    var e;
    return ((e = this._highlightedCounts) == null ? void 0 : e.get(t.label)) ?? 0;
  }
  _getCountSortValue(t, e) {
    return !(this._config.moveFilteredToTop ?? !1) || !this._highlightedCounts ? t.count : e > 0 ? e : t.count;
  }
  get _maxDisplayedItems() {
    return this._config.maxDisplayedItems ?? M;
  }
  _createSearch() {
    const t = document.createElement("div");
    return t.classList.add("search-container"), this._clearButton = document.createElement("button"), this._clearButton.classList.add("clear-button"), this._clearButton.innerHTML = "×", this._clearButton.addEventListener("click", this._clearSearch), this._searchInput = document.createElement("input"), this._searchInput.classList.add("search-input"), this._searchInput.placeholder = "Search...", this._searchInput.addEventListener("input", this._handleSearch), t.appendChild(this._searchInput), t.appendChild(this._clearButton), t;
  }
  _updateButtonState(t, e) {
    t.classList.add("selected"), t.classList.remove("unselected"), e.classList.add("unselected"), e.classList.remove("selected");
  }
  _updateArrowText(t, e, i) {
    t.textContent = `${e ? "↑" : "↓"}${i}`;
  }
  _createSortingHandlers(t, e, i, s) {
    return {
      alpha: () => {
        this._config.sort === "alphabetical" && (i = !i, this._updateArrowText(t, i, "ABC")), this._config.sort = "alphabetical", this._config.sortAscending = i, this._updateButtonState(t, e), this._handleSort();
      },
      numeric: () => {
        this._config.sort === "count" && (s = !s, this._updateArrowText(e, s, "123")), this._config.sort = "count", this._config.sortAscending = s, this._updateButtonState(e, t), this._handleSort();
      },
      reset: () => {
        var o, a;
        (a = (o = this._config).onClick) == null || a.call(o, void 0), this.setSelectedItem(void 0);
      }
    };
  }
  _createSortingBlock() {
    const t = document.createElement("div"), e = document.createElement("div"), [i, s] = [document.createElement("div"), document.createElement("div")], o = document.createElement("div");
    t.classList.add("sorting-block"), e.classList.add("buttons-container"), o.classList.add("reset-text", "unselected");
    const a = this._config.sort === "alphabetical" ? this._config.sortAscending : !0, h = this._config.sort === "count" ? this._config.sortAscending : !0;
    return this._updateArrowText(i, a ?? !1, "ABC"), this._updateArrowText(s, h ?? !1, "123"), o.textContent = "reset selection", this._config.sort && this._updateButtonState(this._config.sort === "alphabetical" ? i : s, this._config.sort === "alphabetical" ? s : i), this._sortingHandlers = this._createSortingHandlers(i, s, a ?? !1, h ?? !1), i.addEventListener("click", this._sortingHandlers.alpha), s.addEventListener("click", this._sortingHandlers.numeric), o.addEventListener("click", this._sortingHandlers.reset), e.append(i, s), t.append(e, o), t;
  }
  _createShowMoreButton() {
    const t = document.createElement("div");
    t.classList.add("show-more"), t.addEventListener("click", this._handleShowMore);
    const e = document.createElement("span");
    return this._showMoreCount = document.createElement("span"), this._showMoreCount.classList.add("show-more-count"), t.appendChild(e), t.appendChild(this._showMoreCount), t;
  }
  _calculateVisibleItems() {
    const t = this._itemHeight + this._itemMargin;
    this._visibleItemsCount = Math.ceil(this._scrollContainer.clientHeight / t);
  }
  _createBarElement(t, e = !1, i) {
    var _, d, g, C, S, v, y;
    if (e) {
      const c = document.createElement("div");
      c.classList.add("bar-highlighted"), c.dataset.label = t.label, c.style.position = "absolute";
      const f = i / t.count * (t.count / this._maxCount) * 100;
      return c.style.width = `${f}%`, c;
    }
    const s = document.createElement("div");
    if (s.classList.add("row"), s.dataset.label = t.label.toString(), s.style.position = "absolute", (_ = this._config) != null && _.onClick && (s.style.cursor = "pointer", s.addEventListener("click", (c) => {
      var f, I;
      return (I = (f = this._config).onClick) == null ? void 0 : I.call(f, t, c);
    })), this._selectedLabels.size) {
      const c = this._container.querySelector(".reset-text");
      c == null || c.classList.remove("unselected"), c == null || c.classList.add("selected"), s.classList.add(this._selectedLabels.has(t.label) ? "selected" : "unselected");
    }
    const o = t.count / this._maxCount * 100, a = document.createElement("div");
    a.classList.add("bar"), a.style.width = `${o}%`;
    const h = document.createElement("div");
    h.classList.add("info");
    const r = document.createElement("div");
    r.classList.add("label"), r.textContent = t.label, r.title = t.label;
    const l = document.createElement("div");
    if (l.classList.add("count"), this._isFilteringActive && i !== void 0) {
      const c = ((g = (d = this._config).countFormatter) == null ? void 0 : g.call(d, i)) ?? i.toString();
      if (this._config.showTotalWhenFiltered) {
        const f = ((S = (C = this._config).countFormatter) == null ? void 0 : S.call(C, t.count)) ?? t.count.toString();
        l.textContent = `${c} / ${f}`;
      } else
        l.textContent = c;
    } else
      l.textContent = ((y = (v = this._config).countFormatter) == null ? void 0 : y.call(v, t.count)) ?? t.count.toString();
    return h.append(r, l), s.append(a, h), s;
  }
  _renderItems() {
    this._renderItemsToContainer(this._barsContainer, !1), this._renderItemsToContainer(this._highlightedBarsContainer, !0);
  }
  _renderItemsToContainer(t, e) {
    var s;
    const i = this._itemHeight + this._itemMargin;
    if (e) {
      if (!this._highlightedCounts) {
        t.replaceChildren();
        return;
      }
      const o = /* @__PURE__ */ new Map();
      for (const h of Array.from(t.children)) {
        const r = h;
        r.classList.contains("bar-highlighted") && r.dataset.label && o.set(r.dataset.label, r);
      }
      const a = /* @__PURE__ */ new Set();
      for (let h = this._startIndex; h < this._endIndex; h++) {
        const r = this._data[h];
        if (!r)
          continue;
        const l = this._highlightedCounts.get(r.label) || 0;
        if (l > 0) {
          a.add(r.label);
          const _ = o.get(r.label);
          if (_) {
            const d = l / r.count * (r.count / this._maxCount) * 100;
            _.style.width = `${d}%`, _.style.top = `${h * i}px`;
          } else {
            const d = this._createBarElement(r, !0, l);
            d.style.top = `${h * i}px`, d.style.height = `${this._itemHeight}px`, t.appendChild(d);
          }
        }
      }
      for (const [h, r] of o)
        a.has(h) || r.remove();
    } else {
      const o = document.createDocumentFragment();
      for (let a = this._startIndex; a < this._endIndex; a++) {
        const h = this._data[a];
        if (!h)
          continue;
        const r = ((s = this._highlightedCounts) == null ? void 0 : s.get(h.label)) || 0, l = this._createBarElement(h, !1, r);
        l.style.top = `${a * i}px`, l.style.height = `${this._itemHeight}px`, o.appendChild(l);
      }
      t.replaceChildren(o);
    }
  }
  _updateContainerHeights(t) {
    const e = Math.min(this._originalData.length, this._maxDisplayedItems) * t;
    this._scrollContainer.style.height = `${e}px`;
    const i = `${this._data.length * t}px`;
    this._barsContainer.style.height = i, this._highlightedBarsContainer.style.height = i;
  }
  _updateShowMoreButton() {
    if (this._originalData.length > this._maxDisplayedItems) {
      const t = this._data.length > this._maxDisplayedItems;
      this._showMoreButton.style.visibility = t ? "visible" : "hidden", this._showMoreButton.style.display = "flex";
    } else
      this._showMoreButton.style.display = "none";
  }
  _updateUIElements(t = p.showSortingBlock, e = p.showSearch) {
    this._originalData.length ? this._data.length ? (this._sortingBlock.style.visibility = t ? "visible" : "hidden", this._searchContainer.style.display = e ? "block" : "none", this.hideState(), this._scrollContainer.style.visibility = "visible") : (this._sortingBlock.style.visibility = t ? "visible" : "hidden", this._searchContainer.style.display = e ? "block" : "none", this._scrollContainer.style.visibility = "visible", this.showState(this._searchInput.value.length ? "No results found" : void 0)) : (this._sortingBlock.style.visibility = "hidden", this._searchContainer.style.display = "none", this._scrollContainer.style.visibility = "hidden", this._showMoreButton.style.display = "none", this.showState());
  }
  _render() {
    this._maxCount = Math.max(...this._data.map((s) => s.count), 1), this._calculateItemDimensions();
    const { showSortingBlock: t, showSearch: e } = this._config, i = this._itemHeight + this._itemMargin;
    this._updateContainerHeights(i), this._updateShowMoreButton(), this._updateUIElements(t, e), this._calculateVisibleItems(), this._updateVisibleItems();
  }
  _calculateItemDimensions() {
    if (!this._data.length)
      return;
    const t = this._createBarElement(this._data[0]);
    this._barsContainer.appendChild(t);
    const e = window.getComputedStyle(t);
    this._itemHeight = t.offsetHeight, this._itemMargin = parseFloat(e.marginTop) + parseFloat(e.marginBottom), this._barsContainer.removeChild(t);
  }
  _calcHiddenSum() {
    if (this._data.length > this._maxDisplayedItems) {
      const t = this._data.slice(this._maxDisplayedItems);
      this._isFilteringActive ? this._hiddenItemsSum = t.reduce((e, i) => {
        var o;
        const s = ((o = this._highlightedCounts) == null ? void 0 : o.get(i.label)) || 0;
        return e + s;
      }, 0) : this._hiddenItemsSum = t.reduce((e, i) => e + i.count, 0);
    } else
      this._hiddenItemsSum = 0;
    this._handleShowMoreCount();
  }
  /**
   * Returns the current configuration of the bars module.
   * @returns {BarsConfigInterface} The current configuration.
   */
  getConfig() {
    return this._config;
  }
  /**
   * Sets the configuration for the bars module.
   * @param {BarsConfigInterface} [config] - The configuration object to set.
   * @returns {void}
   */
  setConfig(t) {
    const e = this._config;
    if (this._config = w(p, t ?? {}), this.setDisplayStateConfig(b.extractConfig(this._config, p)), this._isExpanded = this._config.expanded ?? !1, e.sort !== this._config.sort || e.sortAscending !== this._config.sortAscending || e.moveFilteredToTop !== this._config.moveFilteredToTop) {
      const s = this._isSortAnimationEnabled() ? this._captureSortPositions() : void 0;
      this._isConfigResortInProgress = !0, this._handleSort(), this._isConfigResortInProgress = !1, s && this._animateSortFrom(s);
      return;
    }
    this._calcHiddenSum(), this._render();
  }
  /**
   * Sets the data for the bars module.
   * @param {any[] | undefined} [values] - The array of values to set as the data.
   * @returns {void}
   */
  setData(t = []) {
    var i, s;
    this._clearSearch(), this._uniqueValues = new Set(t), this._countMap = new Int32Array(this._uniqueValues.size);
    const e = new Map([...this._uniqueValues].map((o, a) => [o, a]));
    t.forEach((o) => {
      const a = e.get(o);
      a !== void 0 && this._countMap[a]++;
    }), this._data = [...this._uniqueValues].map((o, a) => ({
      label: o,
      count: this._countMap[a] || 0
    })), this._originalData = [...this._data], !this._originalData.length && this._selectedLabels.size && (this.setSelectedItems(void 0), (s = (i = this._config).onClick) == null || s.call(i, void 0)), this._searchContainer.style.display = this._originalData.length ? "block" : "none", this._handleSort();
  }
  /**
   * Sorts the data in the bars module by the count of each item, in ascending or descending order.
   * @param {boolean} [ascending=false] - If true, sorts the data in ascending order by count. Otherwise, sorts in descending order.
   * @returns {void}
   */
  sortByCount(t = !1) {
    const e = !this._isConfigResortInProgress && this._isSortAnimationEnabled() ? this._captureSortPositions() : void 0;
    this._data.sort((i, s) => {
      const o = this._getFilteredCount(i), a = this._getFilteredCount(s);
      if ((this._config.moveFilteredToTop ?? !1) && this._isFilteringActive && this._highlightedCounts) {
        const d = o > 0, g = a > 0;
        if (d !== g)
          return d ? -1 : 1;
      }
      const h = this._getCountSortValue(i, o), r = this._getCountSortValue(s, a), l = t ? h - r : r - h;
      if (l)
        return l;
      const _ = t ? i.count - s.count : s.count - i.count;
      return _ || i.label.toString().localeCompare(s.label.toString());
    }), this._calcHiddenSum(), this._render(), e && this._animateSortFrom(e);
  }
  /**
   * Sorts the data in the bars module alphabetically by the label of each item, in ascending or descending order.
   * @param {boolean} [ascending=true] - If true, sorts the data in ascending alphabetical order by label. Otherwise, sorts in descending alphabetical order.
   * @returns {void}
   */
  sortAlphabetically(t = !0) {
    const e = !this._isConfigResortInProgress && this._isSortAnimationEnabled() ? this._captureSortPositions() : void 0;
    this._data.sort((i, s) => {
      if ((this._config.moveFilteredToTop ?? !1) && this._isFilteringActive && this._highlightedCounts) {
        const h = this._getFilteredCount(i) > 0, r = this._getFilteredCount(s) > 0;
        if (h !== r)
          return h ? -1 : 1;
      }
      const o = t ? i.label.toString().localeCompare(s.label.toString()) : s.label.toString().localeCompare(i.label.toString());
      if (o)
        return o;
      const a = t ? i.count - s.count : s.count - i.count;
      return a || 0;
    }), this._calcHiddenSum(), this._render(), e && this._animateSortFrom(e);
  }
  /**
   * Destroys the bars module by removing event listeners, clearing the container, and resetting internal data structures.
   * This method should be called when the bars module is no longer needed to clean up any resources it has allocated.
   */
  destroy() {
    this._scrollContainer.removeEventListener("scroll", this._handleScroll), this._searchInput.removeEventListener("input", this._handleSearch), this._clearButton.removeEventListener("click", this._clearSearch), this._showMoreButton.removeEventListener("click", this._handleShowMore);
    const t = this._sortingBlock.querySelector("div:first-child"), e = this._sortingBlock.querySelector("div:nth-child(2)"), i = this._sortingBlock.querySelector(".reset-text");
    t == null || t.removeEventListener("click", this._sortingHandlers.alpha), e == null || e.removeEventListener("click", this._sortingHandlers.numeric), i == null || i.removeEventListener("click", this._sortingHandlers.reset), this._container.innerHTML = "", this._data = [], this._originalData = [], this._uniqueValues.clear(), this._countMap = new Int32Array(0), this._hiddenItemsSum = 0, this._highlightedCounts = void 0, this._isFilteringActive = !1, super.destroy();
  }
  resetList() {
    this._scrollContainer.scrollTop = 0, this._searchInput.value = "", this._selectedItem = void 0, this._selectedLabels.clear(), this._highlightedCounts = void 0, this._isFilteringActive = !1, this._data = [...this._originalData], this._updateVisibleItems();
  }
  /**
   * Gets the currently selected item in the bars module.
   * @returns {BarData | undefined} - The currently selected bar data item, or undefined if none selected.
   */
  getSelectedItem() {
    return this._selectedItem;
  }
  getSelectedItems() {
    return this._selectedLabels.size ? this._originalData.filter((t) => this._selectedLabels.has(t.label)) : [];
  }
  /**
   * Sets the currently selected item in the bars module.
   * @param {BarData | undefined} item - The bar data item to select, or undefined to clear the selection.
   * @returns {void}
   */
  setSelectedItem(t) {
    this.setSelectedItems(t ? [t] : void 0);
  }
  setSelectedItems(t) {
    if (this._selectedLabels.clear(), t != null && t.length)
      for (const e of t)
        this._selectedLabels.add(e.label);
    if (this._selectedItem = (t == null ? void 0 : t.length) === 1 ? t[0] : void 0, !this._selectedLabels.size) {
      const e = this._container.querySelector(".reset-text");
      e == null || e.classList.remove("selected"), e == null || e.classList.add("unselected");
    }
    this._updateVisibleItems();
  }
  /**
   * Sets the highlighted data for crossfiltering visualization.
   * Creates an overlay bar showing the proportion of filtered data for each category.
   * @param data - Array of string labels that should be highlighted, or undefined to clear highlighting
   */
  setHighlightedData(t) {
    if (t == null)
      this._highlightedCounts = void 0, this._isFilteringActive = !1;
    else {
      this._highlightedCounts = /* @__PURE__ */ new Map();
      for (const i of t)
        this._highlightedCounts.set(i, (this._highlightedCounts.get(i) || 0) + 1);
      const e = this._originalData.reduce((i, s) => i + s.count, 0);
      this._isFilteringActive = t.length < e;
    }
    this._handleSort();
  }
  /**
   * Sets the loading state of the bars module by updating the text content of a fallback div.
   * This method is typically called when data is being loaded asynchronously to provide a visual indication to the user.
   */
  setLoadingState() {
    super.setLoadingState();
  }
}
export {
  $ as Bars
};
