// @interacta/css-labels/dist/css-label.js
import { doRectsIntersect } from './helper.js';
import { LEFT_RIGHT_PADDING, TOP_BOTTOM_PADDING, DEFAULT_FONT_SIZE } from './variables.js';
import { cssLabelStyles, injectStyles, labelClassName, hiddenLabelClassName } from './styles.js';
let globalCssLabelStyles;
export class CssLabel {
    /**
     * @param container - The parent element for the label.
     * @param text - Initial label content (plain text or, if dangerousHtml is true, HTML).
     * @param dontInjectStyles - When true, global styles are not injected.
     * @param dangerousHtml - When true, text is set via innerHTML (XSS risk). Only use with trusted/sanitized content.
     */
    constructor(container, text, dontInjectStyles, dangerousHtml) {
        this.element = document.createElement('div');
        this.fontWidthHeightRatio = 0.6;
        this._x = 0;
        this._y = 0;
        /** Real size from getBoundingClientRect; set only after appendChild when not yet set; cleared when size options change. */
        this._cachedRealWidth = undefined;
        this._cachedRealHeight = undefined;
        /** Heuristic size (font + padding); updated in _measureText when _needsMeasureUpdate. */
        this._estimatedWidth = 0;
        this._estimatedHeight = 0;
        this._visible = false;
        this._prevVisible = false;
        this._weight = 0;
        this._needsMeasureUpdate = true;
        this._customFontSize = undefined;
        this._customColor = undefined;
        this._customOpacity = undefined;
        this._shouldBeShown = false;
        this._text = '';
        /**
         * Tracks whether content was set via `dangerouslySetHtml` (`true`) or `setText` (`false`).
         * Needed so that switching mode with the same string (e.g. `setText` then `dangerouslySetHtml`) still updates the DOM.
         */
        this._contentIsHtml = false;
        this._customPadding = undefined;
        this._rotation = 0;
        if (!dontInjectStyles && !globalCssLabelStyles)
            globalCssLabelStyles = injectStyles(cssLabelStyles);
        this._container = container;
        this._updateClasses();
        if (text !== undefined) {
            if (dangerousHtml) {
                this.dangerouslySetHtml(text);
            }
            else {
                this.setText(text);
            }
        }
        this.resetFontSize();
        this.resetPadding();
    }
    /**
     * Width used for layout/overlap: real (getBoundingClientRect) when available, otherwise estimated.
     */
    get width() {
        var _a;
        this._measureText();
        return (_a = this._cachedRealWidth) !== null && _a !== void 0 ? _a : this._estimatedWidth;
    }
    /**
     * Height used for layout/overlap: real (getBoundingClientRect) when available, otherwise estimated.
     */
    get height() {
        var _a;
        this._measureText();
        return (_a = this._cachedRealHeight) !== null && _a !== void 0 ? _a : this._estimatedHeight;
    }
    /**
     * Sets the text of the element using textContent (safe from XSS).
     * @param text - The text to set.
     */
    setText(text) {
        if (this._text !== text || this._contentIsHtml) {
            this._text = text;
            this._contentIsHtml = false;
            this.element.textContent = typeof text === 'number' ? String(text) : text;
            this._needsMeasureUpdate = true;
            this._resetRealSizeCache();
        }
    }
    /**
     * Sets the inner HTML of the element. Only use with trusted or sanitized content.
     * WARNING: XSS risk — do not pass user-provided or unsanitized HTML.
     * @param html - The HTML to set (string or number; numbers are converted to string).
     */
    dangerouslySetHtml(html) {
        if (this._text !== html || !this._contentIsHtml) {
            this._text = html;
            this._contentIsHtml = true;
            this.element.innerHTML = typeof html === 'number' ? String(html) : html;
            this._needsMeasureUpdate = true;
            this._resetRealSizeCache();
        }
    }
    /**
     * Sets the position of the label
     * @param x - The x coordinate of the label
     * @param y - The y coordinate of the label
     */
    setPosition(x, y) {
        this._x = x;
        this._y = y;
    }
    /**
     * Sets the CSS style of the element.
     * If a color, opacity or pointer-events is specified using the `setColor`,
     * `setOpacity` or `setPointerEvents` method, it takes priority over all custom styles.
     * The `fontSize` style will not apply from `setStyle`, and the `transform` style
     * will not apply, as it is used in the draw method to update the label position.
     * @param style - The style to be set.
     */
    setStyle(style) {
        if (this._customStyle !== style) {
            this._customStyle = style;
            this.element.style.cssText = this._customStyle;
            if (this._customColor)
                this.element.style.color = this._customColor;
            if (this._customOpacity)
                this.element.style.opacity = String(this._customOpacity);
            if (this._customPointerEvents)
                this.element.style.pointerEvents = this._customPointerEvents;
            if (this._customFontSize)
                this.element.style.fontSize = `${this._customFontSize}px`;
            if (this._customPadding) {
                const { top, right, bottom, left } = this._customPadding;
                this.element.style.padding = `${top}px ${right}px ${bottom}px ${left}px`;
            }
            this._resetRealSizeCache();
        }
    }
    /**
     * Sets the rotation of the label in degrees. Positive = clockwise.
     * @param rotation - Degrees (0 = horizontal).
     */
    setRotation(rotation) {
        if (this._rotation !== rotation) {
            this._rotation = rotation;
            this._needsMeasureUpdate = true;
            this._resetRealSizeCache();
        }
    }
    /**
     * Resets the label rotation to 0 (horizontal).
     */
    resetRotation() {
        if (this._rotation !== 0) {
            this._rotation = 0;
            this._needsMeasureUpdate = true;
            this._resetRealSizeCache();
        }
    }
    /**
     * Sets the class name of the component
     * @param className - The class name to be set
     */
    setClassName(className) {
        if (this._customClassName !== className) {
            this._customClassName = className;
            this._updateClasses();
            this._resetRealSizeCache();
        }
    }
    /**
     * Sets the font size of the text in pixels.
     * This value cannot be changed through `setStyle` or `setClassName`
     * methods because it is used to measure the width and height of the label.
     * @param fontSize - The font size to set. If not specified, it will use the default value of `14px`.
     */
    setFontSize(fontSize = DEFAULT_FONT_SIZE) {
        if (this._customFontSize !== fontSize) {
            this.element.style.fontSize = `${fontSize}px`;
            this._customFontSize = fontSize;
            this._needsMeasureUpdate = true;
            this._resetRealSizeCache();
        }
    }
    /**
     * Resets the font size of the element to default value.
     */
    resetFontSize() {
        if (this._customFontSize !== DEFAULT_FONT_SIZE) {
            this.element.style.fontSize = `${DEFAULT_FONT_SIZE}px`;
            this._customFontSize = DEFAULT_FONT_SIZE;
            this._needsMeasureUpdate = true;
            this._resetRealSizeCache();
        }
    }
    /**
     * Sets the color of the element.
     * This color will rewrite the color from `setStyle` CSS style if specified.
     * @param color - The color to set
     */
    setColor(color) {
        if (this._customColor !== color) {
            this.element.style.color = color;
            this._customColor = color;
        }
    }
    /**
     * Resets the color of the element.
     */
    resetColor() {
        this.element.style.removeProperty('color');
        this._customColor = undefined;
    }
    /**
     * Sets the opacity of the element.
     * This opacity will rewrite the opacity from `setStyle` CSS style if specified.
     * @param opacity - The opacity to set.
     */
    setOpacity(opacity) {
        if (this._customOpacity !== opacity) {
            this.element.style.opacity = String(opacity);
            this._customOpacity = opacity;
        }
    }
    /**
     * Resets the opacity of the element.
     */
    resetOpacity() {
        this.element.style.removeProperty('opacity');
        this._customOpacity = undefined;
    }
    /**
     * Sets the `pointerEvents` property to 'none', 'auto', or 'all'.
     * This `pointerEvents` value will rewrite the opacity from `setStyle` CSS style if specified.
     * @param pointerEvents - The `pointerEvents` value to be set.
     */
    setPointerEvents(pointerEvents) {
        if (this._customPointerEvents !== pointerEvents) {
            this.element.style.pointerEvents = `${pointerEvents}`;
            this._customPointerEvents = pointerEvents;
        }
    }
    /**
     * Resets the pointer-events of the element.
     */
    resetPointerEvents() {
        this.element.style.removeProperty('pointer-events');
        this._customPointerEvents = undefined;
    }
    /**
     * Sets the padding of the element in pixels.
     * This value cannot be changed through `setStyle` or `setClassName`
     * methods because it is used to measure the width and height of the label.
     * @param padding - The padding object with left, top, right and bottom properties.
     * If not specified or partially specified, it will use the default value of
     * `{ left: 9px, top: 6px, right: 9px, bottom: 6px }` for unspecified values.
     */
    setPadding(padding = {
        left: LEFT_RIGHT_PADDING,
        top: TOP_BOTTOM_PADDING,
        right: LEFT_RIGHT_PADDING,
        bottom: TOP_BOTTOM_PADDING,
    }) {
        if (!this._customPadding ||
            this._customPadding.left !== padding.left ||
            this._customPadding.top !== padding.top ||
            this._customPadding.right !== padding.right ||
            this._customPadding.bottom !== padding.bottom) {
            this._customPadding = padding;
            this.element.style.padding = `${padding.top}px ${padding.right}px ${padding.bottom}px ${padding.left}px`;
            this._needsMeasureUpdate = true;
            this._resetRealSizeCache();
        }
    }
    resetPadding() {
        if (this._customPadding !== undefined &&
            this._customPadding.left === LEFT_RIGHT_PADDING &&
            this._customPadding.top === TOP_BOTTOM_PADDING &&
            this._customPadding.right === LEFT_RIGHT_PADDING &&
            this._customPadding.bottom === TOP_BOTTOM_PADDING) {
            return;
        }
        const padding = {
            left: LEFT_RIGHT_PADDING,
            top: TOP_BOTTOM_PADDING,
            right: LEFT_RIGHT_PADDING,
            bottom: TOP_BOTTOM_PADDING,
        };
        this.element.style.padding = `${padding.top}px ${padding.right}px ${padding.bottom}px ${padding.left}px`;
        this._customPadding = padding;
        this._needsMeasureUpdate = true;
        this._resetRealSizeCache();
    }
    /**
     * Sets the boolean value of whether the element should be forced to shown or not
     * @param shouldBeShown - The boolean value to set
     */
    setForceShow(shouldBeShown) {
        this._shouldBeShown = shouldBeShown;
    }
    /**
     * Gets the boolean value of whether the element should be shown or not.
     * @returns The boolean value of whether the element should be shown or not.
     */
    getForceShow() {
        return this._shouldBeShown;
    }
    /**
     * Draws the element to the container and updates the label's coordinate.
     * The label's coordinate updates using `transform` style. It rewrite
     * the `transform` from `setStyle` CSS style if specified.
     */
    draw() {
        const isVisible = this.getVisibility();
        if (isVisible !== this._prevVisible) {
            if (this._prevVisible === false) {
                this._container.appendChild(this.element);
            }
            else {
                this._container.removeChild(this.element);
            }
            this._updateClasses();
            this._prevVisible = isVisible;
        }
        if (isVisible) {
            const rotate = this._rotation !== 0 ? ` rotate(${this._rotation}deg)` : '';
            // When rotated, pivot around the label’s bottom-center so it stays anchored at (x, y).
            if (this._rotation !== 0) {
                this.element.style.transformOrigin = '50% 100%';
            }
            else {
                this.element.style.removeProperty('transform-origin');
            }
            this.element.style.transform = `
        translate(-50%, -100%)
        translate3d(${this._x}px, ${this._y}px, 0)${rotate}
      `;
            // Re-measure after transforms so cache reflects current rotation/state (e.g. after setText/setRotation while visible).
            if (this._cachedRealWidth === undefined || this._cachedRealHeight === undefined) {
                this._updateRealSizeCache();
            }
        }
    }
    overlaps(label) {
        // Use the same box as getLeft/getRight/getTop/getBottom: centered at (_x, _y), bottom at _y
        return doRectsIntersect({
            x: this._x - this.width / 2,
            y: this._y - this.height,
            width: this.width,
            height: this.height,
        }, {
            x: label._x - label.width / 2,
            y: label._y - label.height,
            width: label.width,
            height: label.height,
        });
    }
    setVisibility(visible = true) {
        this._visible = visible;
    }
    getVisibility() {
        return this._visible && (typeof this._text === 'number' || this._text.trim().length > 0);
    }
    isOnScreen(containerWidth, containerHeight) {
        const width = containerWidth !== null && containerWidth !== void 0 ? containerWidth : this._container.offsetWidth;
        const height = containerHeight !== null && containerHeight !== void 0 ? containerHeight : this._container.offsetHeight;
        return this._x > 0 && this._y > 0 && this._x < width && this._y < height;
    }
    setWeight(weight) {
        this._weight = weight;
    }
    getWeight() {
        return this._weight;
    }
    getPrevVisible() {
        return this._prevVisible;
    }
    /**
     * Gets the left edge of the label's bounding box.
     * @returns The x coordinate of the left edge.
     */
    getLeft() {
        return this._x - this.width / 2;
    }
    /**
     * Gets the right edge of the label's bounding box.
     * @returns The x coordinate of the right edge.
     */
    getRight() {
        return this._x + this.width / 2;
    }
    /**
     * Gets the top edge of the label's bounding box.
     * @returns The y coordinate of the top edge.
     */
    getTop() {
        return this._y - this.height;
    }
    /**
     * Gets the bottom edge of the label's bounding box.
     * @returns The y coordinate of the bottom edge.
     */
    getBottom() {
        return this._y;
    }
    /**
     * Appends the element to the top of the container
     */
    raise() {
        this._container.appendChild(this.element);
    }
    /**
     * Removes the element from the DOM.
     */
    destroy() {
        this.element.remove();
    }
    _updateClasses() {
        const isVisible = this.getVisibility();
        if (isVisible) {
            window.requestAnimationFrame(() => {
                this.element.className = `${labelClassName} ${this._customClassName || ''}`;
            });
        }
        else {
            this.element.className = `${labelClassName} ${this._customClassName || ''} ${hiddenLabelClassName}`;
        }
    }
    /** Clears the real-size cache so it is re-measured after next appendChild. */
    _resetRealSizeCache() {
        this._cachedRealWidth = undefined;
        this._cachedRealHeight = undefined;
    }
    /** Fills real-size cache from getBoundingClientRect when element is in DOM. Called after appendChild when cache does not exist. */
    _updateRealSizeCache() {
        if (this.element.parentElement === null)
            return;
        const rect = this.element.getBoundingClientRect();
        this._cachedRealWidth = rect.width;
        this._cachedRealHeight = rect.height;
    }
    /**
     * Updates estimated size when needed. Real size is cached separately after appendChild.
     */
    _measureText() {
        var _a, _b, _c;
        if (!this._needsMeasureUpdate)
            return;
        const { left, top, right, bottom } = (_a = this._customPadding) !== null && _a !== void 0 ? _a : {
            left: LEFT_RIGHT_PADDING,
            top: TOP_BOTTOM_PADDING,
            right: LEFT_RIGHT_PADDING,
            bottom: TOP_BOTTOM_PADDING,
        };
        const fontSize = (_b = this._customFontSize) !== null && _b !== void 0 ? _b : DEFAULT_FONT_SIZE;
        this._estimatedWidth = fontSize * this.fontWidthHeightRatio * ((_c = this.element.textContent) !== null && _c !== void 0 ? _c : '').length + left + right;
        this._estimatedHeight = fontSize + top + bottom;
        this._needsMeasureUpdate = false;
    }
}
