// @blueprintjs/table/lib/cjs/interactions/menus/menuContext.js
"use strict";
/*
 * Copyright 2016 Palantir Technologies, Inc. All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MenuContextImpl = void 0;
var regions_1 = require("../../regions");
var MenuContextImpl = /** @class */ (function () {
    function MenuContextImpl(target, selectedRegions, numRows, numCols) {
        this.target = target;
        this.selectedRegions = selectedRegions;
        this.numRows = numRows;
        this.numCols = numCols;
        this.regions = regions_1.Regions.overlapsRegion(selectedRegions, target) ? selectedRegions : [target];
    }
    MenuContextImpl.prototype.getTarget = function () {
        return this.target;
    };
    MenuContextImpl.prototype.getSelectedRegions = function () {
        return this.selectedRegions;
    };
    MenuContextImpl.prototype.getRegions = function () {
        return this.regions;
    };
    MenuContextImpl.prototype.getUniqueCells = function () {
        return regions_1.Regions.enumerateUniqueCells(this.regions, this.numRows, this.numCols);
    };
    return MenuContextImpl;
}());
exports.MenuContextImpl = MenuContextImpl;
